Added custom messages from Confluence

colinli02 opened the pull request:
https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1764623016/SOFT+Integration+Planning#Custom-Messages 

<!-- Remember to add the relevant reviewers, assignees, and labels, and create this PR as a draft if it is a work in progress. -->

### Description
<!-- Replace _issue_ with the issue number that this PR resolves, or delete the line and add this PR to the Software project if there is no related issue. -->
- There is no open issue.
- Refer to items here for "Colin" https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/edit-v2/1798504659
<!-- Describe what was done in this PR if there is no related issue, or the related issue's description is not sufficient. -->


### Verification
<!-- List the steps that were taken to verify that the changes introduced by this PR function as desired and without side effects. -->
- Built in ROS (sailbot_workspace) and passed.
![image](https://user-images.githubusercontent.com/102767231/232176227-c82deca1-0a09-4616-9644-ac9deb2c3525.png)

patrick-5546 reviewed (dismissed):
Please add autogenerated files/directories to a `.gitignore` file loll

colinli02 committed:
Added custom ROS messages based on https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/edit-v2/1798504659

patrick-5546 reviewed (commented):
I think we should make this our working spec, to replace https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1764623016/SOFT+Integration+Planning#Interfaces. This way we don't have to update the confluence page during development when we want to make a change to a msg. To do so, we should comment descriptions for fields that are not obvious: conventions, range, etc..
- `speed` and `NUM_WIND_SENSORS` are examples of fields that are obvious
- `direction_degrees` and `voltage` are examples of fields that are not obvious

@hhenry01 this is why I was thinking it would be good to have `Speed` and `Heading` helper messages: we would only have to write the conventions in one place, rather than every time it is used (GPS, Ship, Wind).

patrick-5546 commented on msg/LatLon.msg:
Let's put internal/helper msgs like `LatLon` in a subdirectory

colinli02 replied:
Added helpermsg/ within msg e.g.:
```
msg/helpermsg/LatLon.msg
```

colinli02 committed:
Added new ROS msgs
Moved helper ROS msgs within /helpermsgs
Added comments for helper messages or non-obvious fields

colinli02 commented:
Made changes today. See:
https://github.com/UBCSailbot/ros_msgs/tree/remotes/origin/user/colinli02/custom_messages

Verified by passing build.

colinli02 commented:
### In case anyone sees the dismissed review above:

I dismissed an old review message about adding a .gitignore that was based on an old commit that has been rebased out of log. Had some auto-generated files from an initial creation of tutorial_interfaces (now ros_msgs). These files don't appear anymore so it's all good.

patrick-5546 reviewed (commented):
On the confluence page, I see that the comments about having uppercase field names are unresolved:

- @hhenry01 can we change them?
- @colinli02 please update the confluence page according to Henry's response

Meeting update: get rid of constant fields, fix size of array

~Also Colin just before you merge in this PR, can you publish a new version of the confluence page with a version comment that contains a link to this PR or the PR number?~ Instead, let's stop using the confluence page once this PR is merged in. This repository will serve as the spec for our interfaces. Please remind me to archive the page once this PR is merged in.

patrick-5546 commented on msg/LPathStats.msg:
Let's delete this file for now, we can create it once we know what its fields are

colinli02 replied:
Deleted

patrick-5546 commented on srv/Example.srv:
Let's delete the srv directory, it only contains an example file

colinli02 replied:
/srv deleted

patrick-5546 commented on msg/Ship.msg:
```diff
@@ -0,0 +1,4 @@
+int32 id 
+float32 lat_lon 
+float32 speed 
+float32 heading 
```
Shouldn't the speed and heading fields be `helpermsg` types?

colinli02 replied:
Fixed all to be helper msg type e.g.
```
LatLon lat_lon
Speed speed
Heading heading
```

patrick-5546 commented on msg/helpermsg/Vector3.msg:
I'm thinking this repository should only contain msgs that are used across ROS packages (interfaces). If they are used only in one package, it should be defined in its `msg/` directory. Also maybe we should rename this repository `interfaces`. @colinli02 can you create a discussion topic in next software wide meeting to talk about this (and any questions you have about my review)?

patrick-5546 replied:
@DFriend01

colinli02 replied:
# To-do for me (based on today 2023-06-1 meeting)
- [x] Rename ros_msgs -> ros_interface
- [x] ~~Directly call helper message (try helpermsg/Speed)~~ See comments from last week, we get rid of /helpermsg

patrick-5546 replied:
Can rename repository `custom_interfaces` instead of `ros_interface`. ~Also, do you want to rename the repository? It would involve renaming this GitHub repository and updating `new_project.repos` and the workspace file in sailbot workspace accordingly similar to what you did you add the `ros_msgs` repo.~ I renamed the repository to unblock myself on writing the local pathfinding skeleton code.

patrick-5546 commented:
Just discovered this repository: https://github.com/ros2/common_interfaces/tree/humble

It defines some commonly used messages, including
- sensors, including GPS
- geometry, including vectors
- std: int, float, etc.

It may be worth taking a look at what the ros library offers before create our own custom types. And even if we can't use them directly we can draw inspiration from them.

I commented the lines that are required to add the `geometry_msgs` dependency as an example.

patrick-5546 committed:
Rename package to custom_interfaces and clean up

patrick-5546 committed:
Add tests workflow

patrick-5546 committed:
Add markdownlint config file

patrick-5546 commented on msg/helpermsg/Heading.msg:
~Just realized that subdirectories of `msg/` and `srv` aren't supported: https://github.com/ros2/rosidl/issues/213.~ Nvm they are, but the interface names don't reflect its path. I think we should match that in this repo: get rid of `helpermsg/` and move the helper messages to `msg/`.

This can be verified through the `ros interface show` command:
![image](https://github.com/UBCSailbot/custom_interfaces/assets/65756895/85c58a43-2e94-4c15-ae25-bd6aa65f1f44)

patrick-5546 replied:
You can also try to find standard library msgs instead of defining our own helper messages: https://github.com/UBCSailbot/custom_interfaces/pull/2#issuecomment-1577866615

colinli02 replied:
I've moved from ```helpermsg/``` to ```msg/``` . For some reason that command doesnt work (ros2 interface show custom_interfaces/)
![image](https://github.com/UBCSailbot/custom_interfaces/assets/102767231/a09d1c9c-4f5f-412b-a20b-93e41f135535)

So i manually verified using ```ros2 interface show <msg_type>```
![image](https://github.com/UBCSailbot/custom_interfaces/assets/102767231/2d1436d3-5b4f-4aa4-b8bd-35152c065c9c)

patrick-5546 replied:
That wasn't a command, it was autocomplete (press tab)

patrick-5546 replied:
I think it would still be good to have some way to distinguish interfaces and their helpers if possible, here are my initial thoughts:

1. ~Do msgs have to be capitalized? What if we make interfaces uppercase and helpers lowercase?~ Not too sure about this, looking around I don't see anywhere where people have made lowercase messages. In fact, that may be confusing: https://answers.ros.org/question/328729/ros-msg-definitions-difference-between-std_msgstype-and-type/?answer=328730#post-id-328730

2. Prefix interfaces with `Inter`? And/or prefix helpers with `Helper`?

patrick-5546 commented on msg/Path.msg:
Can you rename this to `DesiredHeading.msg`

colinli02 replied:
re-named

patrick-5546 commented on msg/AIS.msg:
Rename to AISShips.msg

colinli02 replied:
Renamed

patrick-5546 commented on msg/Ship.msg:
Rename to AISShip.msg

colinli02 replied:
Renamed

patrick-5546 commented on msg/Wind.msg:
Rename to WindSensors.msg

colinli02 replied:
Renamed

patrick-5546 reviewed (commented):
More renaming

patrick-5546 committed:
Update tests inputs

patrick-5546 committed:
Use tests in sailbot workspace main branch

patrick-5546 committed:
Update CI to v1.2.1

patrick-5546 committed:
Configure dependabot

colinli02 commented:
Future to-do: look at existing libraries https://github.com/UBCSailbot/custom_interfaces/pull/2#issuecomment-1577866615

colinli02 commented:
# EDIT: Resolved. I forgot confluence page != PR 
# See bottom of comment

Also super weird, I don't know what's causing this but I noticed this last time and this time, ```Ship.msg``` must exist or else ```custom_interfaces``` will not build. 

See error message:
```Executing task in folder sailbot_workspace: ./build.sh 

Starting >>> custom_interfaces
Starting >>> network_systems
Finished <<< network_systems [0.16s]                                                                  
--- stderr: custom_interfaces                            
In file included from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_c/custom_interfaces/msg/detail/ais_ships__functions.h:19,
                 from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_c/custom_interfaces/msg/detail/ais_ships__functions.c:4:
/workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_c/custom_interfaces/msg/detail/ais_ships__struct.h:22:10: fatal error: custom_interfaces/msg/detail/ship__struct.h: No such file or directory
   22 | #include "custom_interfaces/msg/detail/ship__struct.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/custom_interfaces__rosidl_generator_c.dir/build.make:231: CMakeFiles/custom_interfaces__rosidl_generator_c.dir/rosidl_generator_c/custom_interfaces/msg/detail/ais_ships__functions.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:213: CMakeFiles/custom_interfaces__rosidl_generator_c.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
In file included from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_typesupport_cpp/custom_interfaces/msg/ais_ships__type_support.cpp:7:
/workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_cpp/custom_interfaces/msg/detail/ais_ships__struct.hpp:20:10: fatal error: custom_interfaces/msg/detail/ship__struct.hpp: No such file or directory
   20 | #include "custom_interfaces/msg/detail/ship__struct.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_typesupport_fastrtps_cpp/custom_interfaces/msg/detail/ais_ships__rosidl_typesupport_fastrtps_cpp.hpp:11,
                 from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_typesupport_fastrtps_cpp/custom_interfaces/msg/detail/dds_fastrtps/ais_ships__type_support.cpp:4:
/workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_cpp/custom_interfaces/msg/detail/ais_ships__struct.hpp:20:10: fatal error: custom_interfaces/msg/detail/ship__struct.hpp: No such file or directory
   20 | #include "custom_interfaces/msg/detail/ship__struct.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/custom_interfaces__rosidl_typesupport_cpp.dir/build.make:128: CMakeFiles/custom_interfaces__rosidl_typesupport_cpp.dir/rosidl_typesupport_cpp/custom_interfaces/msg/ais_ships__type_support.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:395: CMakeFiles/custom_interfaces__rosidl_typesupport_cpp.dir/all] Error 2
gmake[2]: *** [CMakeFiles/custom_interfaces__rosidl_typesupport_fastrtps_cpp.dir/build.make:154: CMakeFiles/custom_interfaces__rosidl_typesupport_fastrtps_cpp.dir/rosidl_typesupport_fastrtps_cpp/custom_interfaces/msg/detail/dds_fastrtps/ais_ships__type_support.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
In file included from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_typesupport_introspection_cpp/custom_interfaces/msg/detail/ais_ships__type_support.cpp:12:
/workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_cpp/custom_interfaces/msg/detail/ais_ships__struct.hpp:20:10: fatal error: custom_interfaces/msg/detail/ship__struct.hpp: No such file or directory
   20 | #include "custom_interfaces/msg/detail/ship__struct.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/custom_interfaces__rosidl_typesupport_introspection_cpp.dir/build.make:154: CMakeFiles/custom_interfaces__rosidl_typesupport_introspection_cpp.dir/rosidl_typesupport_introspection_cpp/custom_interfaces/msg/detail/ais_ships__type_support.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:369: CMakeFiles/custom_interfaces__rosidl_typesupport_introspection_cpp.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:343: CMakeFiles/custom_interfaces__rosidl_typesupport_fastrtps_cpp.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< custom_interfaces [1.36s, exited with code 2]

Summary: 1 package finished [1.62s]
  1 package failed: custom_interfaces
  1 package had stderr output: custom_interfaces
  ```
  
  Resolved by making an empty "Ship.msg". Why? Note this doesn't happen to any other .msg file.
  I made sure to remove Ship.msg from the cmake file, even cleared the build file, still this happens.
  
  Temporary solution was to add an empty Ship.msg so it builds for now.
  
  # Edit resolved
  - Forgot PR comments diverged from confluence documentation
  - Solution: ```Ship[] ships -> AISShip[] ships```

colinli02 committed:
Updated according to PR comments:
- Re-naming of ros msgs .msg
- Deleted /helpermsg, /srv, moved helper msgs to /msg
- Removed not used .msg according to PR comments
- Fixed issue of not calling helper message
- Note issue: Won't build without Ship.msg. Left empty Ship.msg for now

colinli02 committed:
Fixed issue of empty Ship.msg required for build.
- Forgot confluence page diverged from PR

patrick-5546 reviewed (commented):
Some more to dos:
- [x] Verify with @DFriend01 which msgs are internal to boat simulator and delete those (or create PR in boat_simulator to add those)
**Notes: Confirmed that ```Vector3``` and ```Kinematics``` are internal to boat simulator** - CL
- [x] Add comment to describe every field that isn't a custom message (e.g., `uint8`, `int32`)
- [x] Think of ways to distinguish interface and helper msgs and discuss with @patrick-5546 
**(Edit discussion 2023-06-16: Prefix ```Helper``` only, rest don't change) - CL**

- [x] Add to README the name, publishers, and subscribers of each interface/topic
- [x] After PR is merged, create PR in sailbot workspace to switch custom_interfaces branch/version to `main` in `src/new_project.repos`
- [x] Create issue for https://github.com/UBCSailbot/custom_interfaces/pull/2#issuecomment-1590200460

colinli02 committed:
Removed Vector3 and Kinematics
- Since internal to boat simulator
Added comment to describe every field that isn't a custom message
Added Helper to helper messages as prefix
Added to README the name, publishers, and subscribers of each interface

patrick-5546 committed:
Update helper msg names in msgs

patrick-5546 committed:
Update CI version

patrick-5546 committed:
Standardize naming

patrick-5546 committed:
Update README accordingly

patrick-5546 committed:
Proofread comments

patrick-5546 committed:
Simplify static array msgs

patrick-5546 committed:
Fix internal interface names

patrick-5546 reviewed (approved):
Great work on this PR
